README.rst
====================
BME280 Python Driver
====================

.. image:: https://img.shields.io/travis/kbrownlees/bme280.svg
        :target: https://travis-ci.org/kbrownlees/bme280

.. image:: https://img.shields.io/pypi/v/bme280.svg
        :target: https://pypi.python.org/pypi/bme280


Python Driver for the BME280 Temperature/Pressure/Humidity Sensor from Bosch.

* Free software: BSD license

Full credit to https://github.com/IDCFChannel/bme280-meshblu-py for most of the logic, I simply packaged
/ tidied it.


Features
--------

* i2c reading of the bme280 Temperature/Pressure/Humidity sensor
* munin plugins for graphing the results


Installation (Package)
----------------------

For I2C access you must have the 'smbus' package available - for debian based systems install python-smbus.
If you wish to compile it, the python package is smbus-cffi (https://pypi.python.org/pypi/smbus-cffi/)

Then::

        pip install bme280

Unless you have setup a virtualenv you may need to use sudo. Also, if your smbus is installed globally then
you need to use sudo or ensure your virtualenv has access to the global site packages.

Usage
-----

Adafruit have a good run through of setting up their break out version of the bme280 at
https://learn.adafruit.com/adafruit-bme280-humidity-barometric-pressure-temperature-sensor-breakout/wiring-and-test

You will need to know the i2c address being used by the bme280, it is usually 0x76 or 0x77. To verify which
devices are connected you can use::

    i2cdetect -y 1

Then::

    $ read_bme280 --help

    usage: read_bme280 [-h] [--pressure] [--humidity] [--temperature]
                       [--i2c-address I2C_ADDRESS] [--i2c-bus I2C_BUS]

    optional arguments:
      -h, --help            show this help message and exit
      --pressure
      --humidity
      --temperature
      --i2c-address I2C_ADDRESS
      --i2c-bus I2C_BUS

Example::

    $ read_bme280 --i2c-address 0x77

    1017.58 hPa
      50.55 ％
      19.03 ℃

Munin
-----

Three plugins are available in the /munin folder. To use them link them into /etc/munin/plugins::

    ln -s /path/to/bme280/munin* /etc/munin/plugins

You can configure the plugins by editing /etc/munin/plugin-conf.d/munin-node and adding::

    [bme280_*]
    group i2c
    env.I2C_ADDRESS 0x77

You can test it with::

    sudo munin-run bme280_humidity

If you get an error like::

   /etc/munin/plugins/bme280_humidity: 21: /etc/munin/plugins/bme280_humidity: read_bme280: not found

Then Add the follow lines to /etc/munin/plugin-conf.d/munin-node under the [bme280_*] section you
added before (assuming 'which read_bme280' results in /usr/local/bin/read_bme280)::

    env.PATH /usr/local/bin:/usr/bin:/bin

Restart your node and the new graphs should turn up in about 10 minutes::

    sudo /etc/init.d/munin-node restart

Or you can force run munin a couple of times and they should turn up::

    sudo su - munin munin-cron --shell=/bin/bash



HISTORY.rst
# 0.7 (2023-11-29)

* Fix #6 by JLAC11 (thanks!) - Update bme280.py to leave bme280_i2c unexposed



tests/test_bme280.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
test_bme280
----------------------------------

Tests for `bme280` module.
"""

import unittest

from bme280 import bme280


class TestImport(unittest.TestCase):

    def test_import(self):
        pass


bme280/bme280_i2c.py
# -*- coding: utf-8 -*-

import warnings

try:
    import smbus
except ImportError:
    warnings.warn('smbus is not available, you will not be able to use i2c', ImportWarning)
    smbus = None

default_i2c_address = None
default_bus = None


def set_default_i2c_address(i2c_address):
    global default_i2c_address
    default_i2c_address = i2c_address


def set_default_bus(bus_number):
    global default_bus
    if smbus is None:
        raise RuntimeError('smbus is not available, please ensure it is installed correctly (see README)')

    default_bus = smbus.SMBus(bus_number)


def read_byte_data(cmd, bus=None, i2c_address=None):
    if bus is None:
        bus = default_bus
    if i2c_address is None:
        i2c_address = default_i2c_address
    return bus.read_byte_data(i2c_address, cmd)


def write_byte_data(cmd, value, bus=None, i2c_address=None):
    if bus is None:
        bus = default_bus
    if i2c_address is None:
        i2c_address = default_i2c_address
    return bus.write_byte_data(i2c_address, cmd, value)


bme280/bme280.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
Full credit to https://github.com/IDCFChannel/bme280-meshblu-py (or who ever
originally wrote bme280_sample.py)
"""

import argparse
from collections import namedtuple

from . import bme280_i2c

setup_run = False

calibration_h = []
calibration_p = []
calibration_t = []

t_fine = 0.0

Data = namedtuple('Data', ['humidity', 'pressure', 'temperature'])


def reset_calibration():
    global calibration_h, calibration_p, calibration_t, t_fine
    calibration_h = []
    calibration_p = []
    calibration_t = []
    t_fine = 0.0


def populate_calibration_data():
    raw_data = []

    for i in range(0x88, 0x88 + 24):
        raw_data.append(bme280_i2c.read_byte_data(i))
    raw_data.append(bme280_i2c.read_byte_data(0xA1))
    for i in range(0xE1, 0xE1 + 7):
        raw_data.append(bme280_i2c.read_byte_data(i))

    calibration_t.append((raw_data[1] << 8) | raw_data[0])
    calibration_t.append((raw_data[3] << 8) | raw_data[2])
    calibration_t.append((raw_data[5] << 8) | raw_data[4])
    calibration_p.append((raw_data[7] << 8) | raw_data[6])
    calibration_p.append((raw_data[9] << 8) | raw_data[8])
    calibration_p.append((raw_data[11] << 8) | raw_data[10])
    calibration_p.append((raw_data[13] << 8) | raw_data[12])
    calibration_p.append((raw_data[15] << 8) | raw_data[14])
    calibration_p.append((raw_data[17] << 8) | raw_data[16])
    calibration_p.append((raw_data[19] << 8) | raw_data[18])
    calibration_p.append((raw_data[21] << 8) | raw_data[20])
    calibration_p.append((raw_data[23] << 8) | raw_data[22])
    calibration_h.append(raw_data[24])
    calibration_h.append((raw_data[26] << 8) | raw_data[25])
    calibration_h.append(raw_data[27])
    calibration_h.append((raw_data[28] << 4) | (0x0F & raw_data[29]))
    calibration_h.append((raw_data[30] << 4) | ((raw_data[29] >> 4) & 0x0F))
    calibration_h.append(raw_data[31])

    for i in range(1, 2):
        if calibration_t[i] & 0x8000:
            calibration_t[i] = (-calibration_t[i] ^ 0xFFFF) + 1

    for i in range(1, 8):
        if calibration_p[i] & 0x8000:
            calibration_p[i] = (-calibration_p[i] ^ 0xFFFF) + 1

    for i in range(0, 6):
        if calibration_h[i] & 0x8000:
            calibration_h[i] = (-calibration_h[i] ^ 0xFFFF) + 1


def read_adc():
    data = []
    for i in range(0xF7, 0xF7 + 8):
        data.append(bme280_i2c.read_byte_data(i))
    pres_raw = (data[0] << 12) | (data[1] << 4) | (data[2] >> 4)
    temp_raw = (data[3] << 12) | (data[4] << 4) | (data[5] >> 4)
    hum_raw = (data[6] << 8) | data[7]

    return Data(hum_raw, pres_raw, temp_raw)


def read_all():
    data = read_adc()
    return Data(
        read_humidity(data),
        read_pressure(data),
        read_temperature(data)
    )


def read_humidity(data=None):
    if data is None:
        data = read_adc()

    # We need a temperature reading to calculate humidity
    read_temperature(data)
    return compensate_humidity(data.humidity)


def read_pressure(data=None):
    if data is None:
        data = read_adc()

    # We need a temperature reading to calculate pressure
    read_temperature(data)
    return compensate_pressure(data.pressure)


def read_temperature(data=None):
    if data is None:
        data = read_adc()

    return compensate_temperature(data.temperature)


def compensate_pressure(adc_p):
    v1 = (t_fine / 2.0) - 64000.0
    v2 = (((v1 / 4.0) * (v1 / 4.0)) / 2048) * calibration_p[5]
    v2 += ((v1 * calibration_p[4]) * 2.0)
    v2 = (v2 / 4.0) + (calibration_p[3] * 65536.0)
    v1 = (((calibration_p[2] * (((v1 / 4.0) * (v1 / 4.0)) / 8192)) / 8) + ((calibration_p[1] * v1) / 2.0)) / 262144
    v1 = ((32768 + v1) * calibration_p[0]) / 32768

    if v1 == 0:
        return 0

    pressure = ((1048576 - adc_p) - (v2 / 4096)) * 3125
    if pressure < 0x80000000:
        pressure = (pressure * 2.0) / v1
    else:
        pressure = (pressure / v1) * 2

    v1 = (calibration_p[8] * (((pressure / 8.0) * (pressure / 8.0)) / 8192.0)) / 4096
    v2 = ((pressure / 4.0) * calibration_p[7]) / 8192.0
    pressure += ((v1 + v2 + calibration_p[6]) / 16.0)

    return pressure / 100


def compensate_temperature(adc_t):
    global t_fine
    v1 = (adc_t / 16384.0 - calibration_t[0] / 1024.0) * calibration_t[1]
    v2 = (adc_t / 131072.0 - calibration_t[0] / 8192.0) * (adc_t / 131072.0 - calibration_t[0] / 8192.0) * calibration_t[2]
    t_fine = v1 + v2
    temperature = t_fine / 5120.0
    return temperature


def compensate_humidity(adc_h):
    var_h = t_fine - 76800.0
    if var_h == 0:
        return 0

    var_h = (adc_h - (calibration_h[3] * 64.0 + calibration_h[4] / 16384.0 * var_h)) * (
        calibration_h[1] / 65536.0 * (1.0 + calibration_h[5] / 67108864.0 * var_h * (
            1.0 + calibration_h[2] / 67108864.0 * var_h)))
    var_h *= (1.0 - calibration_h[0] * var_h / 524288.0)

    if var_h > 100.0:
        var_h = 100.0
    elif var_h < 0.0:
        var_h = 0.0

    return var_h


def setup(oversample_t=1, oversample_p=1, oversample_h=1, filter=0, mode=0x3, t_sb=0x5, spi3w_en=0x0):
    """
    Setup BME280 sensor by writing control/configuration registers

    oversample_t - Temperature oversampling factor
    oversample_p - Pressure oversampling factor
    oversample_h - Humidity oversampling factor
    filter       - Sensor filtering factor

    Valid values for these parameters include:
        0 = skip measurement/filter off
        1, 2, 4, 8, or 16

    For remaining parameters, use raw register values, for example:

    mode = 0x3    - Normal power mode
    t_sb = 0x5    - Tstandby 1000ms
    spi3w_en = 0  - 3-wire SPI Disable
    """

    global setup_run
    if setup_run:
        return

    # Oversampling and filter register values (oversampling factor -> register value)
    osrs_values = {
        0: 0x0,
        1: 0x1,
        2: 0x2,
        4: 0x3,
        8: 0x4,
        16: 0x5,
    }
    if oversample_t not in osrs_values:
        raise ValueError("Invalid temperature oversampling value {:d}".format(oversample_t))
    if oversample_p not in osrs_values:
        raise ValueError("Invalid pressure oversampling value {:d}".format(oversample_p))
    if oversample_h not in osrs_values:
        raise ValueError("Invalid humidity oversampling value {:d}".format(oversample_h))
    if filter not in osrs_values:
        raise ValueError("Invalid filter setting {:d}".format(filter))

    osrs_t = osrs_values[oversample_t]
    osrs_p = osrs_values[oversample_p]
    osrs_h = osrs_values[oversample_h]
    filter = osrs_values[filter]

    assert 0x0 <= mode <= 0x3


    ctrl_meas_reg = (osrs_t << 5) | (osrs_p << 2) | mode
    config_reg = (t_sb << 5) | (filter << 2) | spi3w_en
    ctrl_hum_reg = osrs_h

    # The Adafruit Arduino BME280 library notes that:
    # "making sure sensor is in sleep mode before setting configuration as it otherwise may be ignored."
    # https://github.com/adafruit/Adafruit_BME280_Library/blob/e8a7e29df2862109247b7a3eb4f2b10381f4bab3/Adafruit_BME280.cpp#L165
    bme280_i2c.write_byte_data(0xF4, 0x00)

    bme280_i2c.write_byte_data(0xF2, ctrl_hum_reg)
    bme280_i2c.write_byte_data(0xF4, ctrl_meas_reg)
    bme280_i2c.write_byte_data(0xF5, config_reg)

    populate_calibration_data()

    setup_run = True

def full_setup(bus_number,i2c_address,
               oversample_t=1, oversample_p=1, oversample_h=1, filter=0, mode=0x3, t_sb=0x5, spi3w_en=0x0):
    """
    Setup BME280 sensor by writing control/configuration registers

    bus_number   - Default bus number to read data from
    i2c_address  - Default i2c address to read data from
    
    ###
    
    oversample_t - Temperature oversampling factor
    oversample_p - Pressure oversampling factor
    oversample_h - Humidity oversampling factor
    filter       - Sensor filtering factor

    Valid values for these parameters include:
        0 = skip measurement/filter off
        1, 2, 4, 8, or 16

    For remaining parameters, use raw register values, for example:

    mode = 0x3    - Normal power mode
    t_sb = 0x5    - Tstandby 1000ms
    spi3w_en = 0  - 3-wire SPI Disable
    """
    bme280_i2c.set_default_i2c_address(i2c_address)
    bme280_i2c.set_default_bus(bus_number)
    setup(oversample_t,oversample_p,oversample_h,filter,mode,t_sb,spi3w_en)
                   

def main():
    parser = argparse.ArgumentParser()

    parser.add_argument('--pressure', action='store_true', default=False)
    parser.add_argument('--humidity', action='store_true', default=False)
    parser.add_argument('--temperature', action='store_true', default=False)

    parser.add_argument('--i2c-address', default='0x76')
    parser.add_argument('--i2c-bus', default='1')
    args = parser.parse_args()

    if args.i2c_address:
        bme280_i2c.set_default_i2c_address(int(args.i2c_address, 0))
    if args.i2c_bus:
        bme280_i2c.set_default_bus(int(args.i2c_bus))

    setup()
    data_all = read_all()

    if args.pressure:
        print("%7.2f hPa" % data_all.pressure)
    if args.humidity:
        print("%7.2f %%" % data_all.humidity)
    if args.temperature:
        print("%7.2f C" % data_all.temperature)

    if not args.pressure and not args.humidity and not args.temperature:
        print("%7.2f hPa" % data_all.pressure)
        print("%7.2f %%" % data_all.humidity)
        print("%7.2f C" % data_all.temperature)


if __name__ == '__main__':
    main()
